### PR TITLE
[29029] DELETE option for TimeEntries

### PR DESCRIPTION
**Pull request by fupduck**
This is !6833 rebased upon the dev branch.

This adds a DELETE option for TimeEntries in ApiV3.

### CODE REVIEW

**Comment by fupduck** [comment]
I greatly would appreciate help with the tests.

I guess I did not create the project correctly and I have some issues with a wrong content-type.
Sadly I was not able to execute the tests locally - "'development' database is not configured."

File: app/contracts/time_entries/delete_contract.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+module TimeEntries
+  class DeleteContract < BaseContract
```
Comment (ulferts) [inline app/contracts/time_entries/delete_contract.rb:32]: Let's rather inherit from `::ModelContract`. That way, validations defined on the `BaseContract` (e.g. project is the same as the work package's project)  are not taken into account, which seems appropriate given that the time entry is to be deleted anyway.

File: docs/api/apiv3/endpoints/time_entries.apib [hunk 2]
```diff
@@ -107,6 +107,48 @@ Depending on custom fields defined for time entries, additional properties might
...
++ Response 404 (application/hal+json)
+
+    Returned if the time entry does not exist.
+
+    + Body
+
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:NotFound",
+                "message": "The specified user does not exist."
```
Comment (ulferts) [inline docs/api/apiv3/endpoints/time_entries.apib]: While the documentation sometimes claims to return resource specific messages for 404 and 403, the implementation does not. So for 
* 403 => "You are not authorized to access this resource."
* 404 => "The requested resource could not be found."

File: app/services/time_entries/delete_service.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  ##
+  # Deletes the given time entry if allowed.
+  #
+  # @return True if the deletion has been initiated, false otherwise.
+  def call
+    validate_and_yield(time_entry, user) do
+      time_entry.destroy
+    end
```
Comment (ulferts) [inline app/services/time_entries/delete_service.rb:50]: Services are expected to return a `ServiceResult`. So you should change this to:

```
result, errors = validate_and_yield(time_entry, user) do
  time_entry.destroy
end

ServiceResult.new(success: result,
                  errors: errors,
                  result: result)
```

File: lib/api/v3/time_entries/time_entries_api.rb [hunk 4]
```diff
@@ -105,6 +105,14 @@ class TimeEntriesAPI < ::API::OpenProjectAPI
...
               end
             end
+
+            delete do
+              if ::TimeEntries::DeleteService.new(time_entry: @time_entry, user: current_user).call
+                status 202
+              else
+                fail ::API::Errors::Unauthorized
+              end
+            end
```
Comment (ulferts) [inline lib/api/v3/time_entries/time_entries_api.rb:117]: Once the service returns a service result (please see above), this should be changed to:

```
call = ::TimeEntries::DeleteService.new(time_entry: @time_entry, user: current_user).call

if call.success?
  status 202
else
  fail ::API::Errors::ErrorBase.create_and_merge_errors(call.errors)
end
```

That way, the api controller does not need to know why the service call failed, it just knows that it did, and will return an appropriate error based on the errors provided by the service.

File: spec/requests/api/v3/time_entry_resource_spec.rb [hunk 5]
```diff
@@ -493,4 +493,49 @@
       end
     end
   end
+
+  describe 'DELETE /api/v3/time_entries/:id' do
+    let(:path) { api_v3_paths.time_entry time_entry.id }
+
+    before do
+      delete path
```
Comment (ulferts) [inline spec/requests/api/v3/time_entry_resource_spec.rb]: use 

```
delete path, {}, 'CONTENT_TYPE' => 'application/json'
```

to signal the correct content type.

File: spec/requests/api/v3/time_entry_resource_spec.rb [hunk 6]
```diff
@@ -493,4 +493,49 @@
...
+      end
+
+      context 'with the user not being the author' do
+        let(:author) { FactoryBot.create(:user) }
+
+        it_behaves_like 'does not delete the time_entry', 404
+      end
+    end
+  end
+
```
Comment (ulferts) [inline spec/requests/api/v3/time_entry_resource_spec.rb]: remove this line

**Review by ulferts — CHANGES_REQUESTED** [review]
The functionality itself is looking fine 👍 I made comments in the code where I deemed it appropriate.

I didn't have the time to take a closer look at the tests to see what is wrong there.

How are you trying to start the tests? It shouldn't require the `development` database.

**Comment by fupduck** [comment]
Thanks a lot for the review.
Helped a lot to understand what is going on.

I tried running the tests by executing locally
```
$ bash script/ci/setup.sh units postgres
$ bash script/ci/runner.sh units 4 1
```

**Comment by ulferts** [comment]
The ci scripts are not meant to be run locally by a developer.  But it explains the message regarding 'development'. 

You should rather:
* Setup a test environment -> add appropriate statements in config/database.yml
* run the tests. In your case that would mean issuing the following statements in a bash:
  * `rspec ./spec/requests/api/v3/time_entry_resource_spec.rb`
  *  `rspec ./spec/contracts/time_entries/delete_contract_spec.rb`

File: app/services/time_entries/delete_service.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    self.time_entry = time_entry
+    self.contract_class = TimeEntries::DeleteContract
+  end
+
+  ##
+  # Deletes the given time entry if allowed.
+  #
+  # @return True if the deletion has been initiated, false otherwise.
+  def call
+    validate_and_yield(time_entry, user) do
```
Comment (ulferts) [inline app/services/time_entries/delete_service.rb]:
```suggestion
    result, errors = validate_and_yield(time_entry, user) do
```

You need to assign the return values to variables for them to be available later on.

File: app/services/time_entries/delete_service.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      time_entry.destroy
+    end
+
+    ServiceResult.new(success: result, errors: errors, result: result)
```
Comment (ulferts) [inline app/services/time_entries/delete_service.rb]:
```suggestion
    ServiceResult.new(success: result, errors: errors, result: time_entry)
```

My bad, I asked you to provide the wrong variable for the result, sorry.

**Review by ulferts — CHANGES_REQUESTED** [review]
Almost there @fupduck. Keep up the work.

**Comment by ulferts** [comment]
The tests are currently broken on the dev branch but we'll get it working soon. 

The PR itself is looking fine now, so we'll only need Travis' consent now.

**Review by ulferts — APPROVED** [review]

**Comment by fupduck** [comment]
Hey @ulferts  thanks a lot for your help.
I finally (just now) got it running locally and hopefully wlil fix the errors soon.

**Comment by ulferts** [comment]
Glad to hear that @fupduck. I really appreciate your work as OpenProject will benefit from every end point available. With your help, all the basic functionalities on time entries should be available now. The only thing missing now will be telling the client the available values (e.g. on activity) via a form/schema.

**Comment by ulferts** [comment]
To get the test running again on Travis, you'll need to merge the current dev in. Or rebase onto it if that is more to your liking.

**Comment by fupduck** [comment]
@ulferts Now the tests should be alright. Thanks for all the support.

**Comment by ulferts** [comment]
Good job @fupduck :+1:

**Pull request merged**